src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs

```diff
@@ -145,15 +145,12 @@ internal IServiceCallSite GetResolveCallSite(IService service, ISet<Type> callSi
 
         public void Dispose()
         {
-            var disposables = Interlocked.Exchange(ref _disposables, null);
-
-            if (disposables != null)
+            IDisposable disposable;
+            while (_disposables.TryTake(out disposable))
             {
-                foreach (var disposable in disposables)
-                {
-                    disposable.Dispose();
-                }
+                disposable.Dispose();
             }
+            _resolvedServices.Clear();
```

Comment:
This needs a unit test - something where we can prove that that the following works correctly:

- Create a scoped SP
- Resolve a scoped service (instance 1)
- Dispose scoped SP
- Resolve a scoped service (instance 2)
- Assert that 1 and 2 aren't the same instance